torchdp/per_sample_gradient_clip.py

```diff
@@ -195,22 +250,35 @@ def __del__(self):
         self.close()
 
     def close(self):
-        if self.hooks_attached:  # do not close twice
+        if (
+            hasattr(self, "hooks_attached") and self.hooks_attached
+        ):  # do not close twice
             autograd_grad_sample.remove_hooks(self.module)
         self.hooks_attached = False
 
     def __repr__(self):
         return f"PerSampleGradientClipModuleHook on {self.module}"
 
-    def step(self):
-        autograd_grad_sample.compute_grad_sample(self.module, batch_dim=self.batch_dim)
+    def step(self) -> (List[float], int):
+        clip_threshs, batch_size = self.gradient_clipper.clip()
+        return clip_threshs, batch_size
 
-        # The first dim of param.grad_sample is b_sz for every param.
-        # To look up what that value is, we just pick one
-        self.batch_size = next(
-            p.grad_sample.shape[0] for p in self.module.parameters() if p.requires_grad
-        )
+    def virtual_step(self) -> None:
+        """
+        Clips and sums up per-sample gradients into an accumulator.
+        After calling self.virtual_step() N times on mini-batches of
+        B per-sample gradients, a call to self.step() will 
+        populate the .grad field with the average gradient over
+        the entire batch of size N*B.
+        """
+        self.gradient_clipper.clip(virtual=True)
+        # reset the accumulation of per-sample gradients
+        autograd_grad_sample.clear_grad_sample(self.module)
```

Comment:
this kind of goes against having a create_or_extend function unless we allow extension. may be adding a mode like accumulating : bool or sth similar could help. I am not suggesting that we do extension in this diff but knowing it exists help. This way we do not have to delete the grad_sample when simply only accumulating.

Reply:
Note that accumulating gradients in .grad_sample via create_or_extend doesn't require any calls to .virtual_step(). This happens automatically via the backprop hooks.
So my thoughts here were that you can either call .backward() multiple times to have the per-sample gradients accumulate in .grad_sample, or clip and sum up gradients into .summed_grad by calling .virtual_step().
You can also combine these two (there's an example in virtual_step_test.py that does this).

But after you've called .virtual_step(), it is important to clear .grad_sample since those gradients have all been clipped and summed up already, and we wouldn't want to process them again.